### PR TITLE
Introduce bind function based on @yogthos PR

### PR DESCRIPTION
**Pull request by bskinny**
Add a `bind` function which throws LDAPException on failure (for more meaningful error messages) and can return a diagnostic message when password policies are in use.

README still to be completed based on successful code review.

### CODE REVIEW

File: test/clj_ldap/test/client.clj [hunk 1]
```diff
@@ -139,6 +140,18 @@
...
 
+(deftest some-bind
+  (is (= (ldap/bind *conn* (:dn person-a*) "passa")
+         {:code 0, :name "success"}))
+  (is (thrown? LDAPException (ldap/bind *conn* (:dn person-a*) "notthepass")))
+  (is (thrown? LDAPException (ldap/bind *conn* "cn=does,ou=not,cn=exist" "password"))))
+
+(deftest bad-bind?
+  (is (= (ldap/bind? *conn* (:dn person-a*) "not-the-password")
+         false))
```
Comment (sumbach) [inline test/clj_ldap/test/client.clj:151]: I believe `clojure.test/is` assumes the first parameter is the expected value and subsequent parameter(s) are the actual value(s). Test failure messages will make more sense if these test cases are written like:

```
(is (= false (ldap/bind? *conn* (:dn person-a*) "not-the-password")))
```

File: src/clj_ldap/client.clj [hunk 2]
```diff
@@ -524,9 +535,29 @@
...
+bind function will attempt to change the identity of that connection
+to that of the provided DN. Subsequent operations on that connection
+will be done using the bound identity.
+
+If an LDAP connection pool object is passed as the connection argument
+the bind attempt will have no side-effects, leaving the state of the
+underlying connections unchanged."
+  [connection bind-dn password]
+  (let [^BindResult r (basic-bind connection bind-dn password)]
+      (merge (ldap-result r)
```
Comment (sumbach) [inline src/clj_ldap/client.clj:554]: non-standard indentation?

**Comment by sumbach** [comment]
* Overall, this looks great to me -- really nice to expose `bind` like this (I've resorted to calling `.bind` via Java interop in the past).
* Left a couple comments for minor suggestions.
* Not sure how I feel about `basic-bind` as a function name, but this isn't part of the public API so the name isn't critical (nor can I think of anything better to suggest 😉)

**Comment by bskinny** [comment]
@sumbach Thanks for the review! I'll get the issues addressed.

File: src/clj_ldap/client.clj (hunk 2 above)
Comment (bskinny) [reply src/clj_ldap/client.clj:554]: Fixed

File: test/clj_ldap/test/client.clj (hunk 1 above)
Comment (bskinny) [reply test/clj_ldap/test/client.clj]: Went through the test suite and correct.

File: src/clj_ldap/client.clj (hunk 2 above)
Comment (bskinny) [reply src/clj_ldap/client.clj:554]: Fixed

**Review by bskinny — COMMENTED** [review]
PR updated. Trying out the name `bind-based-on-connection` instead of `basic-bind` for the private function.

**Comment by bskinny** [comment]
@pauldorman What do you think of these two commits?
- The first introduces a new `bind` function and is low risk
- The second bumps the unboundid-ldapsdk to the current 5.1.1  release which includes a slew of updates and fixes related to security

**Comment by bskinny** [comment]
@pauldorman Just a nudge to have a look when you can.

File: src/clj_ldap/client.clj (hunk 2 above)
Comment (sumbach) [reply src/clj_ldap/client.clj:554]: Clarification: the indentation on the `(merge ,,,)` form is non-standard.

Comment (bskinny) [reply src/clj_ldap/client.clj]: Ah, thanks @sumbach. Fixed now along with some other indentation alignments.

**Comment by bskinny** [comment]
@sumbach I'll merge this by the end of next week if we do not hear from @pauldorman but Paul will have to deploy to clojars.

**Comment by sumbach** [comment]
Okay by me @bskinny; would be great to update CHANGELOG.md too, since you've bumped the release version in this PR (see #13 for an example).

Thanks again @bskinny and @pauldorman !!

**Comment by bskinny** [comment]
@sumbach Thanks again. CHANGELOG.md updated.

**Comment by bskinny** [comment]
@sumbach @yogthos v0.0.17 is deployed to clojars: [org.clojars.pntblnk/clj-ldap "0.0.17"]

**Pull request merged**